src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -0,0 +1,218 @@
+
+                // Create driver that holds onto compilation and associated analyzers
+                return compilation.WithAnalyzers(analyzers, analysisOptions);
+            }
+
+            private CompilationWithAnalyzersOptions GetAnalyzerOptions(
+                Project project,
+                bool logAnalyzerExecutionTime,
+                bool reportSuppressedDiagnostics)
+            {
+                // in IDE, we always set concurrentAnalysis == false otherwise, we can get into thread starvation due to
+                // async being used with syncronous blocking concurrency.
+                return new CompilationWithAnalyzersOptions(
+                    options: new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Workspace),
+                    onAnalyzerException: GetOnAnalyzerException(project.Id),
+                    analyzerExceptionFilter: GetAnalyzerExceptionFilter(project),
+                    concurrentAnalysis: false,
+                    logAnalyzerExecutionTime: logAnalyzerExecutionTime,
+                    reportSuppressedDiagnostics: reportSuppressedDiagnostics);
+            }
+
+            private Func<Exception, bool> GetAnalyzerExceptionFilter(Project project)
+            {
+                return ex =>
+                {
+                    if (project.Solution.Workspace.Options.GetOption(InternalDiagnosticsOptions.CrashOnAnalyzerException))
+                    {
+                        // if option is on, crash the host to get crash dump.
+                        FatalError.ReportUnlessCanceled(ex);
+                    }
+
+                    return true;
+                };
+            }
+
+            private Action<Exception, DiagnosticAnalyzer, Diagnostic> GetOnAnalyzerException(ProjectId projectId)
+            {
+                return _owner.Owner.GetOnAnalyzerException(projectId, _owner.DiagnosticLogAggregator);
+            }
+
+            private void ResetAnalyzerDriverMap()
+            {
+                // we basically eagarly clear the cache on some known changes
+                // to let CompilationWithAnalyzer go.
+
+                // we create new conditional weak table every time, it turns out 
+                // only way to clear ConditionalWeakTable is re-creating it.
+                // also, conditional weak table has a leak - https://github.com/dotnet/coreclr/issues/665
+                _map = new ConditionalWeakTable<Project, CompilationWithAnalyzers>();
+            }
+
+            [Conditional("DEBUG")]
+            private void AssertAnalyzers(CompilationWithAnalyzers analyzerDriver, IEnumerable<StateSet> stateSets)
+            {
+                if (analyzerDriver == null)
+                {
+                    // this can happen if project doesn't support compilation or no stateSets are given.
+                    return;
+                }
+
+                // make sure analyzers are same.
```

Comment:
HeeJae, you are talking about suppression for individual diagnostic instances (by pragma, suppress message etc.). That should be controlled by the caller. However, if an analyzer is itself disabled (all its rules are effectively disabled, either by default or due to effective severity none), CompilationWithAnalyzers will never run that analyzer - there is no way to get diagnostics for such analyzers. We should ensure we don't pass in such disabled analyzers to CompilationWithAnalyzers ctor. Basically we need to do this: http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs,188